File: gke_customization/gke_catalog/report/sof_by_category,_design_by,_setting_type,_order_type/sof_by_category,_design_by,_setting_type,_order_type.py
# Copyright (c) 2025, Gurukrupa Export and contributors
# For license information, please see license.txt

import frappe

def execute(filters=None):
    columns = get_columns(filters)
    data = get_data(filters)
    return columns, data


def get_columns(filters):
    # group_label = "Category" if filters.get("group_by") == "Category" else "Order Type"
    group_by = filters.get("group_by")

    if group_by == "Category":
        group_label = "Category"
    elif group_by == "Order Type":
        group_label = "Order Type"
    elif group_by == "Design By":
        group_label = "Design By"
    elif group_by == "Setting Type":
        group_label = "Setting Type"
    elif group_by == "Customer Group":
        group_label = "Customer Group"
    else:
        group_label = "Group"

    return [
        {"label": group_label, "fieldname": "label", "fieldtype": "Data", "width": 180},
        {"label": "Total Orders", "fieldname": "total_orders", "fieldtype": "Int", "width": 130},
        {"label": "Submitted", "fieldname": "submitted_orders", "fieldtype": "Int", "width": 130},
        {"label": "Pending", "fieldname": "pending_orders", "fieldtype": "Int", "width": 130},
        {"label": "Cancelled", "fieldname": "cancelled_orders", "fieldtype": "Int", "width": 130},
    ]


def get_data(filters):
    conditions = []
    
    if filters.get("from_date") and filters.get("to_date"):
        conditions.append(f"DATE(skf.creation) BETWEEN '{filters.get('from_date')}' AND '{filters.get('to_date')}'")

    if filters.get("company"):
        conditions.append(f"skf.company = '{filters.get('company')}'")

    if filters.get("branch"):
        conditions.append(f"skf.branch = '{filters.get('branch')}'")

    where_clause = "WHERE " + " AND ".join(conditions) if conditions else ""

    group_by = filters.get("group_by") or "Category"

    if group_by == "Category":
        group_field = "skfd.category"
        join_clause = "LEFT JOIN `tabSketch Order Form Detail` skfd ON skfd.parent = skf.name"
    elif group_by == "Order Type":
        group_field = "skf.order_type"
        join_clause = ""
    elif group_by == "Design By":
        group_field = "skf.design_by"
        join_clause = ""
    elif group_by == "Setting Type":
        group_field = "skfd.setting_type"
        join_clause = "LEFT JOIN `tabSketch Order Form Detail` skfd ON skfd.parent = skf.name"
    elif group_by == "Customer Group":
        join_clause = """
            LEFT JOIN `tabCustomer` ct ON skf.customer_code = ct.name
            LEFT JOIN `tabCustomer Group` cc ON ct.customer_group = cc.name
        """
        select_label = """
            CASE 
                WHEN cc.customer_group_name IN ('Groups', 'Individual', 'Common Party', 'Wholesale') THEN 'Wholesale'
                WHEN cc.customer_group_name = 'Internal' THEN 'Internal'
                WHEN cc.customer_group_name = 'Retail' THEN 'Retail'
                ELSE 'Other-Groups'
            END
        """
        group_field = select_label
    else:
          # fallback
        group_field = "skfd.category"
        join_clause = "LEFT JOIN `tabSketch Order Form Detail` skfd ON skfd.parent = skf.name"

    query = f"""
        SELECT 
            {group_field} AS label,
            COUNT(distinct skf.name) AS total_orders,
            COUNT(DISTINCT CASE WHEN skf.docstatus = 1 THEN skf.name END) AS submitted_orders,
            COUNT(DISTINCT CASE WHEN skf.docstatus = 0 THEN skf.name END) AS pending_orders,
            COUNT(DISTINCT CASE WHEN skf.docstatus = 2 THEN skf.name END) AS cancelled_orders
        FROM `tabSketch Order Form` skf
        {join_clause}
        {where_clause}
        GROUP BY {group_field}
    """

    return frappe.db.sql(query, as_dict=True)

File: gke_customization/gke_catalog/report/sof_by_category,_design_by,_setting_type,_order_type/sof_by_category,_design_by,_setting_type,_order_type.js
// Copyright (c) 2025, Gurukrupa Export and contributors
// For license information, please see license.txt

frappe.query_reports["SOF By Category, Design By, Setting Type, Order Type"] = {
    filters: [
        {
            fieldname: "group_by",
            label: __("Group By"),
            fieldtype: "Select",
            options: ["Category", "Order Type", "Design By", "Setting Type"],  //Customer Group
            default: "Category",
            reqd: 1
        },
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date"
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date"
        },
        // {
        //     fieldname: "company",
        //     label: __("Company"),
        //     fieldtype: "Link",
        //     options: "Company"
        // },
        {
            fieldname: "branch",
            label: __("Branch"),
            fieldtype: "Link",
            options: "Branch"
        }
    ],

    formatter: function (value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        const group_by = frappe.query_report.get_filter_value("group_by");
        const group_value = data.label;
        const from_date = frappe.query_report.get_filter_value("from_date") || "";
        const to_date = frappe.query_report.get_filter_value("to_date") || "";
        const company = frappe.query_report.get_filter_value("company") || "";
        const branch = frappe.query_report.get_filter_value("branch") || "";

        if (["total_orders", "submitted_orders", "pending_orders", "cancelled_orders"].includes(column.fieldname)) {
            
            let status_filter = "";
            if (column.fieldname === "submitted_orders") status_filter = "&docstatus=1";
            else if (column.fieldname === "pending_orders") status_filter = "&docstatus=0";
            else if (column.fieldname === "cancelled_orders") status_filter = "&docstatus=2";

            let url = `/app/sketch-order-form?view=list`
                + (from_date ? `&from_date=${from_date}` : "")
                + (to_date ? `&to_date=${to_date}` : "")
                + (company ? `&company=${company}` : "")
                + (branch ? `&branch=${branch}` : "");

            if (group_by === "Category") url += `&category=${encodeURIComponent(group_value)}`;
            else if (group_by === "Order Type") url += `&order_type=${encodeURIComponent(group_value)}`;
            else if (group_by === "Design By") url += `&design_by=${encodeURIComponent(group_value)}`;
            else if (group_by === "Setting Type") url += `&setting_type=${encodeURIComponent(group_value)}`;
           // else if (group_by === "Customer Group") url += `&customer_code=${encodeURIComponent(group_value)}`;

           // else url += `&order_type=${encodeURIComponent(group_value)}`;
            //design_by

            url += status_filter;

            return `<a href="${url}" target="_blank">${value}</a>`;
        }

        return value;
    }
};

